File: Round 3/recalculating.cpp
// Copyright (c) 2020 kamyu. All rights reserved.

/*
 * Google Code Jam 2020 Round 3 - Problem D. Recalculating
 * https://codingcompetitions.withgoogle.com/codejam/round/000000000019ff7e/00000000003775e9
 *
 * Time:  O(N^2 * logN)
 * Space: O(N^2)
 *
 */

#include<bits/stdc++.h>

using namespace std;

using Groups = unordered_map<int64_t, vector<tuple<int64_t, int64_t, int64_t, int64_t>>>;
using Points = vector<array<int64_t, 2>>;

static const int64_t MOD = 1e9 + 7;
static const int64_t P = 113;

uint64_t pow(uint64_t a, uint64_t b, uint64_t m) {
    a %= m;
    uint64_t result = 1;
    while (b) {
        if (b & 1) {
            result = (result * a) % m;
        }
        a = (a * a) % m;
        b >>= 1;
    }
    return result;
}

static unordered_map<uint64_t, uint64_t> lookup;
int64_t exp_(uint64_t x) {
    if (!lookup.count(x)) {
        lookup[x] = pow(P, x, MOD);
    }
    return lookup[x];
}

uint64_t gcd(uint64_t a, uint64_t b) {
    while (b != 0) {
        const auto tmp = b;
        b = a % b;
        a = tmp;
    }
    return a;
}

pair<uint64_t, Groups> group_rects(const Points& points, int64_t D) {
    unordered_set<int64_t> x_set, y_set;
    for (auto& p : points) {
        x_set.emplace(p[0] - D);
        x_set.emplace(p[0] + D);
        y_set.emplace(p[1] - D);
        y_set.emplace(p[1] + D);
    }
    vector<int64_t> xs(cbegin(x_set), cend(x_set)), ys(cbegin(y_set), cend(y_set));
    sort(begin(xs), end(xs)), sort(begin(ys), end(ys));
    Groups groups;
    uint64_t total = 0;
    for (int j = 0; j < ys.size() - 1; ++j) {
        int64_t rolling_hash = 0;
        deque<int> dq;
        int left = 0, right = 0;
        for (int i = 0; i < xs.size() - 1; ++i) {
            for (; right < points.size() && points[right][0] <= xs[i] + D; ++right) {
                if (ys[j + 1] - D <= points[right][1] &&  points[right][1] <= ys[j] + D) {
                    if (!dq.empty()) {
                        auto a = dq.back();
                        auto x = ((points[right][0] - points[a][0]) % MOD + MOD) % MOD;
                        auto y = ((points[right][1] - points[a][1]) % MOD + MOD) % MOD;
                        rolling_hash = (rolling_hash * P * P + (x * P + y)) % MOD;
                    }
                    dq.emplace_back(right);
                }
            }
            for (; left < points.size() && points[left][0] < xs[i + 1] - D; ++left) {
                if (ys[j + 1] - D <= points[left][1] && points[left][1] <= ys[j] + D) {
                    auto a = dq.front(); dq.pop_front();
                    if (dq.size() >= 1) {
                        auto b = dq.front();
                        auto x = ((points[b][0] - points[a][0]) % MOD + MOD) % MOD;
                        auto y = ((points[b][1] - points[a][1]) % MOD + MOD) % MOD;
                        auto delta = ((x * P + y) * exp_(2 * (dq.size() - 1))) % MOD;
                        rolling_hash = ((rolling_hash - delta) % MOD + MOD) % MOD;
                    }
                }
            }
            if (dq.empty()) {
                continue;
            }
            // the rectangle is fully covered by ordered repair centers in dq,
            // normalized by shift to the first repair center
            int64_t x0 = points[dq.front()][0] - (xs[i + 1] - D), y0 = points[dq.front()][1] - (ys[j + 1] - D);
            int64_t x1 = points[dq.front()][0] - (xs[i] - D), y1 = points[dq.front()][1] - (ys[j] - D);
            total += (x1 - x0) * (y1 - y0);
            groups[rolling_hash].emplace_back(x0, y0, x1, y1);
        }
    }
    return {total, groups};
}

uint64_t calc_unique_area(const Groups& groups) {
    using Event = tuple<int64_t, int64_t, int64_t, int64_t>;
    int64_t unique = 0;
    for (const auto& kvp : groups) {
        vector<Event> intervals;
        for (const auto& rect : kvp.second) {
            int64_t x0, y0, x1, y1;
            tie(x0, y0, x1, y1) = rect;
            intervals.emplace_back(x0, y0, y1, +1);
            intervals.emplace_back(x1, y0, y1, -1);
        }
        sort(begin(intervals), end(intervals));
        unordered_set<int64_t> y_set;
        for (const auto& interval : intervals) {
            int64_t x, y0, y1, v;
            tie(x, y0, y1, v) = interval;
            y_set.emplace(y0);
            y_set.emplace(y1);
        }
        vector<int64_t> ys(cbegin(y_set), cend(y_set));
        sort(begin(ys), end(ys));
        unordered_map<int, int> height_to_idx;
        for (int i = 0; i < ys.size(); ++i) {
            height_to_idx[ys[i]] = i;
        }
        using Node = array<int64_t, 3>;
        vector<Node> tree(2 * (ys.size() - 1));
        auto query = [&](int x) {
            int N = tree.size() / 2;
            if (x >= N) {  // leaf node
                for (int i = 0; i < 2; ++i) {
                    tree[x][i] = (i - tree[x][2] == 0) ? ys[(x - N) + 1] - ys[(x - N)] : 0;
                }
            } else {
                for (int i = 0; i < 2; ++i) {
                    tree[x][i] = (i - tree[x][2] >= 0) ? tree[2 * x][i - tree[x][2]] + tree[2 * x + 1][i - tree[x][2]] : 0;
                }
            }
        };
        auto pull = [&](int x) {
            for (; x; x /= 2) {
                query(x);
            }
        };
        auto update = [&](int l, int r, int v) {
            int N = tree.size() / 2;
            l += N;
            r += N;
            int l0 = l, r0 = r;
            for (; l <= r; l /= 2, r /= 2) {
                if (l & 1) {
                    tree[l][2] += v;
                    query(l);
                    ++l;
                }
                if ((r & 1) == 0) {
                    tree[r][2] += v;
                    query(r);
                    --r;
                }
            }
            pull(l0), pull(r0);
        };
        for (int i = tree.size() - 1; i; --i) {
            query(i);
        }
        for (int i = 0; i < intervals.size() - 1; ++i) {
            int64_t x, y0, y1, v;
            tie(x, y0, y1, v) = intervals[i];
            update(height_to_idx[y0], height_to_idx[y1] - 1, v);
            unique += (get<0>(intervals[i + 1]) - x) * tree[1][1];
        }
    }
    return unique;
}

string recalculating() {
    int N; int64_t D;
    cin >> N >> D;
    Points points;
    for (int i = 0; i < N; ++i) {
        int64_t x, y;
        cin >> x >> y;
        points.push_back({x + y, x - y});
    }
    sort(begin(points), end(points));
    uint64_t total;
    Groups groups;
    tie(total, groups) = group_rects(points, D);
    const auto& unique = calc_unique_area(groups);
    const auto& g = gcd(unique, total);
    return to_string(unique / g) + " " + to_string(total / g);
}

int main() {
    ios_base::sync_with_stdio(false), cin.tie(nullptr);
    int T;
    cin >> T;
    for (int test = 1; test <= T; ++test) {
        cout << "Case #" << test << ": " << recalculating() << '\n';
    }
    return 0;
}

File: Round 3/recalculating.py
# Copyright (c) 2020 kamyu. All rights reserved.
#
# Google Code Jam 2020 Round 3 - Problem D. Recalculating
# https://codingcompetitions.withgoogle.com/codejam/round/000000000019ff7e/00000000003775e9
#
# Time:  O(N^2 * logN), even without MLE, it still faces TLE if N = 1687 in both Python2 / PyPy2, but this solution did pass in C++
# Space: O(N^2), it gets MLE if N = 1687, since memory usage of "(x0, y0, x1, y1) * (4*N^2)" is much greater than 1GB in both Python2 / PyPy2
#

from collections import defaultdict, deque
from functools import partial
from fractions import gcd

class SegmentTree(object):
    def __init__(self, N, update_fn, query_fn):
        self.N = N
        self.update_fn = update_fn
        self.query_fn = query_fn
        # node: [len_of_0_covered, len_of_1_covered, count_of_covered]
        self.tree = [[0, 0, 0] for _ in xrange(2*N)]
        for x in reversed(xrange(1, len(self.tree))):
            self.query_fn(self.tree, x)

    def __apply(self, x, val):
        self.update_fn(self.tree[x], val)
        self.query_fn(self.tree, x)

    def update(self, L, R, val):  # Time: O(logN), Space: O(N) 
        def pull(x):
            while x > 1:
                x //= 2
                self.query_fn(self.tree, x)

        L += self.N
        R += self.N
        L0, R0 = L, R
        while L <= R:
            if L & 1:  # is right child
                self.__apply(L, val)
                L += 1
            if R & 1 == 0:  # is left child
                self.__apply(R, val)
                R -= 1
            L //= 2
            R //= 2
        pull(L0), pull(R0)
    
    def query(self):
        return self.tree[1]

def group_rects(points, D):
    x_set, y_set = set(), set()
    for x, y in points:
        x_set.add(x-D)
        x_set.add(x+D)
        y_set.add(y-D)
        y_set.add(y+D)
    xs, ys = sorted(x_set), sorted(y_set)
    exp = [1]
    for _ in xrange(len(points)-1):
        exp.append(exp[-1]*P*P%MOD)
    total = 0
    groups = defaultdict(list)
    for j in xrange(len(ys)-1):
        rolling_hash, left, right = 0, 0, 0
        dq = deque()
        for i in xrange(len(xs)-1):
            while right < len(points) and points[right][0] <= xs[i]+D:
                if ys[j+1]-D <= points[right][1] <= ys[j]+D:
                    if dq:
                        a, b = dq[-1], right
                        rolling_hash = (rolling_hash*P*P +
                                        (points[b][0]-points[a][0])*P+
                                        (points[b][1]-points[a][1]))%MOD
                    dq.append(right)
                right += 1
            while left < len(points) and points[left][0] < xs[i+1]-D:
                if ys[j+1]-D <= points[left][1] <= ys[j]+D:
                    a = dq.popleft()
                    if len(dq) >= 1:
                        b = dq[0]
                        rolling_hash = (rolling_hash -
                                        ((points[b][0]-points[a][0])*P+
                                         (points[b][1]-points[a][1]))*exp[len(dq)-1])%MOD
                left += 1
            if not dq:
                continue
            # the rectangle is fully covered by ordered repair centers in dq,
            # normalized by being relative to the first repair center
            x0, y0 = xs[i]-points[dq[0]][0], ys[j]-points[dq[0]][1]
            x1, y1 = xs[i+1]-points[dq[0]][0], ys[j+1]-points[dq[0]][1]
            total += (x1-x0)*(y1-y0)
            groups[rolling_hash].append((x0, y0, x1, y1))
    return total, groups

def calc_unique_area(groups):
    def update(x, v):
        x[-1] += v

    def query(ys, tree, x):
        N = len(tree)//2
        if x >= N:  # leaf node
            for i in xrange(2):
                tree[x][i] = ys[(x-N)+1]-ys[(x-N)] if i-tree[x][-1] == 0 else 0
        else:
            for i in xrange(2):
                tree[x][i] = tree[2*x][i-tree[x][-1]] + tree[2*x+1][i-tree[x][-1]] if i-tree[x][-1] >= 0 else 0

    unique = 0
    for rects in groups.itervalues():
        intervals, y_set = [], set()
        for x0, y0, x1, y1 in rects:
            intervals.append((x0, (y0, y1), +1))
            intervals.append((x1, (y0, y1), -1))
            y_set.add(y0)
            y_set.add(y1)
        intervals.sort(key=lambda x: x[0])  # at most O(N^2) intervals, total time: O(N^2 * logN)
        ys = sorted(y_set)
        y_to_idx = {y:i for i, y in enumerate(ys)}
        segment_tree = SegmentTree(len(ys)-1, update_fn=update, query_fn=partial(query, ys))
        for i in xrange(len(intervals)-1):
            x0, (y0, y1), v = intervals[i]
            segment_tree.update(y_to_idx[y0], y_to_idx[y1]-1, v)  # at most O(N^2) intervals, total time: O(N^2 * logN)
            unique += (intervals[i+1][0]-x0)*segment_tree.query()[1]
    return unique

def recalculating():
    N, D = map(int, raw_input().strip().split())
    points = []
    for _ in xrange(N):
        x, y = map(int, raw_input().strip().split())
        points.append((x+y, x-y))
    points.sort()
    total, groups = group_rects(points, D)  # Time: O(N^2)
    unique = calc_unique_area(groups)  # Time: O(N^2 * logN)
    g = gcd(unique, total)
    return "{} {}".format(unique//g, total//g)

MOD = 10**9+7
P = 113
for case in xrange(input()):
    print 'Case #%d: %s' % (case+1, recalculating())
